cpp/dolfinx/la/MatrixCSR.h

```diff
@@ -94,17 +114,79 @@ void add_csr(U&& data, const V& cols, const V& row_ptr, const W& x,
     // Columns indices for row
     auto cit0 = std::next(cols.begin(), row_ptr[row]);
     auto cit1 = std::next(cols.begin(), row_ptr[row + 1]);
-    for (std::size_t c = 0; c < xcols.size(); ++c)
+    for (std::size_t c = 0; c < nc; ++c)
     {
       // Find position of column index
       auto it = std::lower_bound(cit0, cit1, xcols[c]);
       assert(it != cit1);
       std::size_t d = std::distance(cols.begin(), it);
       assert(d < data.size());
-      data[d] += xr[c];
+
+      if constexpr (BS > 0)
+      {
+        for (int i = 0; i < BS; ++i)
+          for (int j = 0; j < BS; ++j)
+            data[d * BS * BS + i * BS + j] += xr[(i * nc + c) * BS + j];
+      }
+      else
+      {
+        for (int i = 0; i < bs[0]; ++i)
+          for (int j = 0; j < bs[1]; ++j)
+            data[d * nbs + i * bs[1] + j] += xr[(i * nc + c) * bs[1] + j];
+      }
+    }
+  }
+}
+
+/// @brief Add block data to a CSR matrix of blocksize 1
+///
+/// @param[out] data The CSR matrix data
+/// @param[in] cols The CSR column indices
+/// @param[in] row_ptr The pointer to the ith row in the CSR data
+/// @param[in] x The `m` by `n` dense block of values (row-major) to add
+/// to the matrix
+/// @param[in] xrows The block row indices of `x`
+/// @param[in] xcols The block column indices of `x`
+/// @param[in] bs Number of entries in each block
+template <typename U, typename V, typename W, typename X>
+void add_csr_blocked(U&& data, const V& cols, const V& row_ptr, const W& x,
```

Comment:
It's not clear how this function relates to `add_csr`, which also has a block size.